python/revng/ptml/tree.py

```diff
@@ -0,0 +1,99 @@
+#
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+# noqa: A003
+
+from dataclasses import dataclass
+from functools import cached_property
+from typing import Dict, Generator, List
+
+
+@dataclass
+class Position:
+    """Represents a single character in a multi-line text document"""
+
+    line: int
+    character: int
+
+    def clone(self):
+        return Position(self.line, self.character)
+
+    def to_string(self):
+        return f"{self.line}:{self.character}"
+
+
+@dataclass
+class Range:
+    """Represents a range of characters in a multi-line text document. Both
+    start and end are inclusive"""
+
+    start: Position
+    end: Position
+
+    def is_single_line(self):
+        return self.start.line == self.end.line
+
+    def to_string(self):
+        return f"{self.start.to_string()},{self.end.to_string()}"
+
+
+class TextExtractor:
+    """Utility function that allows extracting `Range`s from a text"""
+
+    def __init__(self, text: str):
+        self.original_text = text
+
+    @cached_property
+    def lines(self) -> List[str]:
+        return self.original_text.split("\n")
+
+    def get_text(self, range_: Range) -> str:
+        if range_.is_single_line():
+            return self._get_line(range_.start.line, range_.start.character, range_.end.character)
+        lines = [self._get_line(range_.start.line, range_.start.character)]
+        for line_index in range(range_.start.line, range_.end.line + 1):
+            lines.append(self.lines[line_index])
+        lines.append(self._get_line(range_.end.line, -1, range_.end.character))
+        return "\n".join(lines)
+
+    def _get_line(self, index: int, start=-1, end=-1):
+        line = self.lines[index]
+        if start == -1 and end == -1:
+            return line
+        elif start == -1:
+            return line[:end]
+        elif end == -1:
+            return line[start:]
+        else:
+            return line[start:end]
+
+
+@dataclass
+class TreeNode:
+    """The node of a tree of ranged attributes"""
+
+    attributes: Dict[str, str]
+    children: "List[TreeNode]"
```

Comment:
What do you think of `from __future__ import annotations`?

Reply:
I don't remember when, but at some point the target python version was 3.6, which did not have `annotations`.
Given we're targeting 3.8 right now I'll go ahead and use it.

Reply:
Added the import where needed